### PR TITLE
Refactored CRUD Update - PUT and PATCH methods and test for the same

### PR DESCRIPTION
**Pull request by SharanyaT**
DONOT MERGE UNTIL 122 and 124 are merged in
1. Refactored PUT and PATCH methods in payments.py
2. test cases for both in test_payments.py
3. DB handler code for both are in interfaces.py
4. NEED TO WORK ON : is_valid, set default in interfaces.py and write test case for passing invalid payment id

### CODE REVIEW

File: app/db/interface.py [hunk 1]
```diff
@@ -46,8 +57,31 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
...
         :param payment_attributes: <dict> a collection of new payment attribute values that will overwrite old ones
         """
-
-        raise NotImplementedError()
+        
+        payment = Payment.query.get_or_404(payment_id)
+        if payment_replacement:
+            #need to handle is valid and throw 400 bad request for payment_replacement
+            payment.deserialize(payment_replacement)
+            self.db.session.commit()
```
Comment (odoric27) [inline app/db/interface.py]: you need to self.db.session.add(payment) before you commit. otherwise nothing gets saved to db

File: app/db/interface.py [hunk 2]
```diff
@@ -46,8 +57,31 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
...
+            print 'inside interface patch'
+            # implement isvalid on payment_attributes and throw 400 bad request otherwise
+            existing_payment = payment.serialize()
+            payload_nickname = existing_payment['nickname'] if 'nickname' not in payment_attributes else payment_attributes['nickname']
+            payload_type = existing_payment['type'] if 'type' not in payment_attributes else payment_attributes['type']
+            payload_detail = existing_payment['detail'] if 'detail' not in payment_attributes else payment_attributes['detail']
+            payload_default = existing_payment['default'] if 'default' not in payment_attributes else payment_attributes['default'] # set only one default
+            payload_charge = existing_payment['charge-history'] if 'charge-history' not in payment_attributes else payment_attributes['charge-history']
+            target_payment = {'id' : payment_id, 'nickname' : payload_nickname, 'default' : payload_default,
+                'charge-history' :  payload_charge, 'type' : payload_type, 'detail' : payload_detail}
```
Comment (odoric27) [inline app/db/interface.py]: the fields in the model are `charge_history` and `is_default` (note the underscores instead of dashes)

also, this won't find changed inside the details dict. it only checks to see if there is a 'detail' key in `payment_attributes`, but not if a detail field is being changed. as an example, a user will be updating information like {'card_number' : '123'}, but we can't handle that request here.

also , not sure how to implement exactly, but we'd ultimately want to guard against users changing fields like `is_default` and `charge_history` through a regular update. these things should only be modified by the actions explicitly designed for them. food for thought.

File: app/db/interface.py [hunk 3]
```diff
@@ -46,8 +57,31 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
...
+            payment.deserialize(target_payment)
+            self.db.session.commit()
```
Comment (odoric27) [inline app/db/interface.py]: same note about adding to db before commit

File: app/db/interface.py (hunk 1 above)
Comment (SharanyaT) [inline app/db/interface.py]: `@app.route('/pets/<int:id>', methods=['PUT'])
def update_pets(id):
    pet = Pet.query.get_or_404(id)
    pet.deserialize(request.get_json())
    db.session.commit()
    return make_response(jsonify(pet.serialize()), status.HTTP_200_OK)`
Referring to this piece of code from Professor's example, I don't think we would need to do an add since it's an update?

File: app/db/interface.py (hunk 2 above)
Comment (SharanyaT) [inline app/db/interface.py]: 1. Will change the field names - working on it
2. For now, we can only update details field completely.  We can hold off on the feature of updating sub fields of details for the next sprint
3. As for guarding against changing is_default and charge history - Currently writing a is_valid function to make sure such changes don't happen

Maybe we can hold off on merging this PR until I push these changes in

Comment (odoric27) [inline app/db/interface.py]: and a general note: might be easier to construct a loop that only checks the values passed in `payment_attributes`. forgive any syntax errors, but something like that:

```
for key in payment_attributes:
    try:
        existing_payment[key] = payment_attributes[key]
    except KeyError as e:
        raise DataValidationError('Invalid payment: missing ' + e.args[0])
    except TypeError as e:
        raise DataValidationError('Invalid payment: body of request contained bad or no data')
```

File: app/db/interface.py (hunk 1 above)
Comment (odoric27) [inline app/db/interface.py]: ah, you're totally right. my bad.

Comment (SharanyaT) [inline app/db/interface.py]: No issues

File: tests/test_payments.py [hunk 4]
```diff
@@ -109,3 +119,86 @@ def test_crud_create_garbage(self, mock_ps_add):
...
+            mock_ps_update.assert_called_with(mock.ANY, payment_replacement=None)
+            self.assertEqual(resp.status_code, status.HTTP_400_BAD_REQUEST)
+
+# passing garbage data to put
+    @mock.patch.object(PaymentService, 'update_payment', side_effect=DataValidationError, autospec=True)
+    def test_crud_update_patch_garbage(self, mock_ps_update):
+        garbage = 'a@$*&@#sdassdc3r 3284723X43&^@!#@*#'
+        data = json.dumps(garbage)
+        try:
+            resp = self.app.patch('/payments/1', data=data, content_type='application/json')
```
Comment (odoric27) [inline tests/test_payments.py]: these try/except blocks are broken. i had mine like this previously so apologies if that's where you get the idea.

but you'll see that nothing under the except actually runs because that error isn't actually raised by the patch method. it's raised by the mocked update_payment, but then not handled in the caller.

i noticed this in my own code my placing an assert I knew should fail inside that block. when i ran it, everything passed because that test assertion wasn't even made if the exception is caught.

one way to handle it would be creating a context where the assertion is handled in the test:

```
with self.assertRaises(DataValidationError) as context:
    resp = self.app.patch('/payments/1', data=data, content_type='application/json')

self.assertTrue('your custom error message' in context.exception.message)
```

alternatively, you could encapsulate the CRUD update in a try/except and then handled the error accordingly there and assert in the test that however you handled it is done properly.

File: app/db/interface.py [hunk 5]
```diff
@@ -110,15 +136,49 @@ def _query_payments(self, payment_attributes):
...
         except exc.SQLAlchemyError:
             raise PaymentServiceQueryError('Could not retrieve payment items due to query error with given attributes')
 
+# UTILITY FUNCTIONS
+    def is_valid_put(self,old_data,new_data):
+        valid = False
+        valid_detail = False
+        try:
+            if new_data['is_removed'] and new_data['is_removed'] == True :
+                 raise DataValidationError('Invalid payment: Changes to is_removed field not allowed')
```
Comment (odoric27) [inline app/db/interface.py]: why check two conditions here?
`if new_data['is_removed'] and new_data['is_removed'] == True :`

don't we just want to confirm whether or not `old_data` has been removed ?

File: app/error_handlers.py [hunk 6]
```diff
@@ -30,4 +30,20 @@ def method_not_allowed(e):
 
 @app.errorhandler(500)
 def internal_error(e):
-    return make_response(jsonify(status=500, error='Internal Server Error', message='Well, this is embarrassing...'), status.HTTP_500_INTERNAL_SERVER_ERROR)
+	return make_response(jsonify(status=500, error='Internal Server Error', message='Well, this is embarrassing...'), status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+class InvalidPaymentID(Exception):
+    status_code = 404
```
Comment (odoric27) [inline app/error_handlers.py]: there's no test coverage for these functions

also, i don't think we need a status_code class variable if you're setting an instance variable of the same name via passed parameters

Comment (SharanyaT) [inline app/error_handlers.py]: I think i missed to make a comment that test cases for this exception is in progress and will mostly be part of next sprint - taking forever to figure out some mocking errors

File: app/payments.py [hunk 7]
```diff
@@ -126,61 +126,55 @@ def get_payments(id):
...
+            raise DataValidationError('Invalid payment: body of request contained bad or no data')
+        message = payment_service.update_payment(id,payment_replacement=data)
+        rc = HTTP_200_OK
+    except InvalidPaymentID as e:
+        message = 'Invalid payment: Payment ID not found'
         rc = HTTP_404_NOT_FOUND
-
+    except DataValidationError as e:
+        message = 'Invalid payment: body of request contained bad or no data'
+        rc = HTTP_400_BAD_REQUEST
```
Comment (odoric27) [inline app/payments.py]: these error messages can be pulled from the exception by `e.message` instead of rewriting them here

File: app/db/models.py [hunk 8]
```diff
@@ -45,6 +45,27 @@ def deserialize(self, data):
...
+
+    def deserialize_put(self, data):
+        try:
+            self.user_id = data['user_id']
+            self.nickname = data['nickname']
+            self.payment_type = data['payment_type']
+            self.is_default = data['is_default']
+            self.is_removed = False
+            self.charge_history = data['charge_history']
+            details = data['details']
```
Comment (odoric27) [inline app/db/models.py]: no test coverage for deserialize put

File: app/db/interface.py [hunk 9]
```diff
@@ -56,8 +60,29 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
         :param payment_replacement: <dict> a complete payload that describes a new payload which replaces the old one
         :param payment_attributes: <dict> a collection of new payment attribute values that will overwrite old ones
         """
-
-        raise NotImplementedError()
+        payment = self.db.session.query(Payment).get(payment_id)
```
Comment (odoric27) [inline app/db/interface.py]: no test coverage for update

Comment (SharanyaT) [inline app/db/interface.py]: Will a different PR and thats in a new branch

**Comment by odoric27** [comment]
need to re-review this branch as i haven't looked at the last several pushes. if possible, would be best to review once all applicable test cases are ready. don't want to merge something into master that's potentially broken.

but anyone else should feel free to review as well or in place of me

**Comment by odoric27** [comment]
also - if we want to merge this branch in sans `is_valid_put`, might be best to start a new branch for that helper function and its tests just so we can get rid of this behemoth

File: app/db/models.py (hunk 8 above)
Comment (SharanyaT) [inline app/db/models.py]: Implemented in some new commits

File: app/db/interface.py [hunk 10]
```diff
@@ -71,8 +75,30 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
...
+        if payment == None or payment.is_removed == True :
+            raise InvalidPaymentID('Invalid payment: Payment ID not found',status_code=404)
```
Comment (odoric27) [inline app/db/interface.py]: better naming convention would have "Error" at the end of "InvalidPaymentID" just so its easy to see exactly what that is at a glance

File: app/db/interface.py [hunk 11]
```diff
@@ -71,8 +75,30 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
...
+                if key in UPDATABLE_PAYMENT_FIELDS:
+                    existing_payment[key] = payment_attributes[key]
+                else:
+                    raise DataValidationError('Invalid payment: body of request contains invalid/un-updatable fields')
+            payment.deserialize_put(existing_payment)
+            self.db.session.commit()
+            return_object = payment.serialize()
+            return return_object
+        else:
+            raise DataValidationError('Invalid payment: body of request contained bad or no data')
```
Comment (odoric27) [inline app/db/interface.py]: some code repetition here that could be cleaned up. both `if` and `elif` blocks commit, serialize return object and return. we could do all of this once after `else`

File: app/error_handlers.py [hunk 12]
```diff
@@ -30,4 +30,12 @@ def method_not_allowed(e):
 
 @app.errorhandler(500)
 def internal_error(e):
-    return make_response(jsonify(status=500, error='Internal Server Error', message='Well, this is embarrassing...'), status.HTTP_500_INTERNAL_SERVER_ERROR)
+	return make_response(jsonify(status=500, error='Internal Server Error', message='Well, this is embarrassing...'), status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+class InvalidPaymentID(Exception):
+    def __init__(self, message, status_code=None, payload=None):
```
Comment (odoric27) [inline app/error_handlers.py]: same note as above, append 'Error' to classname

Comment (odoric27) [inline app/error_handlers.py]: also - some of the innards of this class might be overkill. i don't see the payload parameter being used anywhere... what were you planning to pass there?

if this error is raised here, it needs to be caught and handled in updates. or we need to designate an error handler to catch these types of errors

File: app/payments.py [hunk 13]
```diff
@@ -126,63 +126,58 @@ def get_payments(id):
...
-                                  'detail' : payload['detail']}
-            message = payments[index[0]]
-            rc = HTTP_200_OK
-        else:
-            message = { 'error' : 'Payments data was not valid' }
-            rc = HTTP_400_BAD_REQUEST
-    else:
-        message = { 'error' : 'Payments %s was not found' % id }
+            raise DataValidationError('Invalid payment: body of request contained bad or no data')
+        try:
```
Comment (odoric27) [inline app/payments.py]: i think we could satisfy both of these `if` blocks by cutting them out and just starting with:

`data = request.get_json(silent=True)`

if request doesn't exist OR if not json, this will "fail silently" meaning it won't raise an exception but simply return None to payload. Then we could either pass `None` to update_payments inside a try/except block and let it handle the error and raise it back if need be.

or let `get_json` throw its error here where we can catch it and do something with it. though i think it would throw a 404 and get picked up by the error handler if nothing else.

specific implementation up to you, but either way, we can cut down on some of this extra code

File: app/payments.py [hunk 14]
```diff
@@ -126,63 +126,58 @@ def get_payments(id):
...
+            data = request.get_json()
+        except Exception:
```
Comment (odoric27) [inline app/payments.py]: same comments as above

File: app/db/interface.py (hunk 9 above)
Comment (odoric27) [inline app/db/interface.py]: this sucker right here (line 78) is what's giving us trouble, though it seems to be a bit more than that.

i was finally able to recreate the db error. this query throws an exception that isn't handled, but it's only throwing it because the tables haven't been built by the db. not sure why so i'm looking into this. just an fyi

File: app/payments.py (hunk 13 above)
Comment (SharanyaT) [inline app/payments.py]: @odoric27 the request.is_json is for when the content type is not json - this cannot be identified and explicitly returned to inform the customer with the get_json call, for eg, text/plain. Also, from what I found when writing the code was get_json does throw a 404 - we need it to be a 400 bad request right? Thats why I handled it this way to raise a 400 instead. And the situation when it is none is handled by the update_payments interface method.

File: app/db/interface.py (hunk 10 above)
Comment (SharanyaT) [inline app/db/interface.py]: @odoric27 Nice catch :) will change that in the next commit! Thanks

File: app/db/interface.py (hunk 11 above)
Comment (SharanyaT) [inline app/db/interface.py]: @odoric27 Noted, will update in the next commit!

File: app/error_handlers.py (hunk 12 above)
Comment (SharanyaT) [inline app/error_handlers.py]: Originally, I was passing the entire request back, but I realized that it was unnecessary, will refactor in next commit

File: app/payments.py (hunk 7 above)
Comment (SharanyaT) [inline app/payments.py]: That does work, but it fails when mocking even if i mock the interface layer's exception with a proper message, will look into it in further detail and make changes

File: app/payments.py (hunk 13 above)
Comment (odoric27) [inline app/payments.py]: that's true. i guess i wasn't too concerned about specifying that in the error message, but makes sense.

curious that it throws a 404. either way, that's why i like the `silent` option, which disables the exception and returns None.

File: app/payments.py (hunk 7 above)
Comment (odoric27) [inline app/payments.py]: yeah, i was having trouble validating error messages with mocks as well. gave up on trying and was content to just assert the right functions were called AND that the functions after an exception was raised were not called

File: app/payments.py (hunk 13 above)
Comment (SharanyaT) [inline app/payments.py]: Ok, so i will retain the is_json check, but will change the request.data check - using silent option, that way the interface method can handle it. Just a question, so when I pass say "sdfdfdfgfgghhhhhhhh324345345345" as the request data, request.get_json will return None, is that correct?

File: app/payments.py (hunk 7 above)
Comment (SharanyaT) [inline app/payments.py]: So do you suggest that I do the same? No issues with changing code, I just want to follow best practices :)

Comment (odoric27) [inline app/payments.py]: at this point, i'm all for the path of least resistance. we can write more tests later to confirm that all of the pieces are working together and the right error messages are being received, but for purposes of mocking in the unit test, i think its sufficient to know that the right code is being touched at the right time

File: app/payments.py (hunk 13 above)
Comment (odoric27) [inline app/payments.py]: yes, i think so

File: app/db/interface.py [hunk 15]
```diff
@@ -71,8 +75,30 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
...
+            #   raise DataValidationError('Invalid payment: body of request contained bad or no data')
+            payment.deserialize_put(payment_replacement)
+            self.db.session.commit()
+            return_object = payment.serialize()
+            return return_object
+        elif payment_attributes: #patch
+            existing_payment = payment.serialize()
+            for key in payment_attributes:
+                if key in UPDATABLE_PAYMENT_FIELDS:
+                    existing_payment[key] = payment_attributes[key]
```
Comment (jdgumz) [inline app/db/interface.py]: Nice!

File: app/error_handlers.py [hunk 16]
```diff
@@ -30,4 +30,12 @@ def method_not_allowed(e):
 
 @app.errorhandler(500)
 def internal_error(e):
-    return make_response(jsonify(status=500, error='Internal Server Error', message='Well, this is embarrassing...'), status.HTTP_500_INTERNAL_SERVER_ERROR)
+	return make_response(jsonify(status=500, error='Internal Server Error', message='Well, this is embarrassing...'), status.HTTP_500_INTERNAL_SERVER_ERROR)
```
Comment (jdgumz) [inline app/error_handlers.py]: Why has this line been indented further by one tab?

File: tests/test_payments.py [hunk 17]
```diff
@@ -273,3 +283,62 @@ def test_index(self):
...
+        mock_ps_update.assert_called_with(mock.ANY,1,payment_attributes=PATCH_CREDIT)
+        self.assertEqual( resp.status_code, status.HTTP_200_OK )
+        new_json = json.loads(resp.data)
+        self.assertEqual (new_json['nickname'], 'boringcredit')
+
+# passing text data to put
+    def test_crud_update_put_with_text_data(self):
+        resp = self.app.put('/payments/1', data="hello", content_type='text/plain')
+        self.assertEqual( resp.status_code, status.HTTP_400_BAD_REQUEST )
+        self.assertTrue('bad or no data' in resp.data)
```
Comment (jdgumz) [inline tests/test_payments.py]: We should probably replace `bad or no data` with an error message constant defined in`payments.py`.  Just import the constant from `payments.py` and use it here.

File: tests/test_payments.py [hunk 18]
```diff
@@ -273,3 +283,62 @@ def test_index(self):
...
+    def test_crud_update_put_garbage(self):
+        garbage = "a@$*&@#sdassdc3r 3284723X43&^@!#@*#"
+        resp = self.app.put('/payments/1',data = "a@$*&@#sdassdc3r 3284723X43&^@!#@*#", content_type='application/json')
+        self.assertTrue('bad or no data' in resp.data)
+        self.assertEqual(resp.status_code, status.HTTP_400_BAD_REQUEST)
+
+# passing garbage data to patch
+    def test_crud_update_patch_garbage(self):
+        garbage = "a@$*&@#sdassdc3r 3284723X43&^@!#@*#"
+        resp = self.app.patch('/payments/1',data = "a@$*&@#sdassdc3r 3284723X43&^@!#@*#", content_type='application/json')
```
Comment (jdgumz) [inline tests/test_payments.py]: Why not `data=garbage`?

File: tests/test_interface.py [hunk 19]
```diff
@@ -399,6 +415,28 @@ def test_interface_delete_payment_with_invalid_id(self, mock_db):
...
         #TODO - next sprint : do get, check count, should be same
 
+#valid PUT data
+    @mock.patch.object(PaymentService,'is_valid_put',return_value = True)
+    @mock.patch.object(app_db, 'session')
+    @mock.patch('app.db.models.Payment')
+    #@mock.patch.object(Payment, 'deserialize_put')
+    #@mock.patch.object(Payment, 'serialize', return_value=PUT_CREDIT_RETURN)
+    def test_interface_update_with_valid_put_data(self,mock_isvalid,mock_db,mock_P):
+        print 'in here'
```
Comment (odoric27) [inline tests/test_interface.py]: here's a working version of this test:

couple important notes:
your sample data, while not the reason for the failure, was incorrect. the field `is_removed` is not returned by serialize.

the sample data `PUT_CREDIT_RETURN` is not an accurate representation of data passed as `payment_replacement`. The user won't know about fields like `charge_history`, `is_default`, `is_removed` so those would need to be updated internally.

serialize is actually called twice (as param in is_valid and at the end of the function)

i think the problem with this code was including the `get(1)`. It works without specifying the params because (i think) we're specify the return result of the `get` function as opposed to the return value of whatever `get(1)` returns. make sense?

```
    @mock.patch.object(PaymentService, 'is_valid_put', return_value=True)
    @mock.patch.object(app_db, 'session')
    @mock.patch('app.db.models.Payment')
    def test_interface_update_with_valid_put_data(self, mock_P, mock_db, mock_is_valid):
        data = PUT_CREDIT

        mock_db.query(Payment).get.return_value = mock_P
        mock_P.serialize.return_value = PUT_CREDIT_RETURN
        resp = self.ps.update_payment(id, payment_replacement=data)
        mock_P.serialize.assert_called()
        mock_is_valid.assert_called_once()
        mock_P.deserialize_put.assert_called_once()
        mock_db.commit.assert_called_once()
        self.assertEqual(resp['nickname'],'favcredit')
```

File: app/db/interface.py [hunk 20]
```diff
@@ -71,8 +75,33 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
...
+        if payment_replacement:
+            print 'jjjjjj'
+            # TODO 	: test cases for validity in next sprint
+            if not self.is_valid_put(payment.serialize(),payment_replacement):
+                print 'not valid'
+                raise DataValidationError('Invalid payment: body of request contained bad or no data')
+            print 'lllllllll'
+            payment.deserialize_put(payment_replacement)
+        elif payment_attributes: #patch
+            existing_payment = payment.serialize()
```
Comment (odoric27) [inline app/db/interface.py]: we would also need to validate `payment_attributes` data in the same way as `payment_replacement`, but more on this below in is_valid function

File: app/db/interface.py [hunk 21]
```diff
@@ -71,8 +75,33 @@ def update_payment(self, payment_id, payment_replacement=None, payment_attribute
...
+        if payment == None or payment.is_removed == True :
+            print 'potato'
+            raise InvalidPaymentID('Invalid payment: Payment ID not found',status_code=404)
+        if not payment_replacement and not payment_attributes:
+            raise DataValidationError('Invalid payment: body of request contained bad or no data')
+        if payment_replacement:
+            print 'jjjjjj'
+            # TODO 	: test cases for validity in next sprint
+            if not self.is_valid_put(payment.serialize(),payment_replacement):
+                print 'not valid'
```
Comment (odoric27) [inline app/db/interface.py]: we shouldn't need the existing payment to confirm if incoming data is valid. more on this in is_valid function

Comment (odoric27) [inline app/db/interface.py]: we shouldn't have to pass in existing data for validation of new data. more on this below

File: app/db/interface.py [hunk 22]
```diff
@@ -135,6 +166,37 @@ def _query_payments(self, payment_attributes):
         except exc.SQLAlchemyError:
             raise PaymentServiceQueryError('Could not retrieve payment items due to query error with given attributes')
 
+# UTILITY FUNCTIONS
+    def is_valid_put(self,old_data,new_data):
+        valid = False
```
Comment (odoric27) [inline app/db/interface.py]: Having trouble understanding why we need to do some of these checks. consider the following:

user provides new car data in json format of full card data similar to that used to create new payments. this data should contain no info about charge_history, default, removed, user_id.

IF that data does contain these fields (or any other illegal fields) then we could catch this by comparing input data against UPDATEABLE_FIELDS. if keys in new data that aren't in UPDATEABLE, throw a bad request error. this could happen before validation or while we build the new payment object itself after validation. If we encounter such a field, we just don't commit the changes and again return a bad request.

the only validation that needs to occur is validation of proper keys/values of the input data, which occurs below those initial checks. in this way we can also remove some perhaps unnecessary code that dampens our coverage. 

does that make sense or am i just not seeing something?

File: tests/test_interface.py (hunk 19 above)
Comment (jdgumz) [inline tests/test_interface.py]: Yeah, I think Derek is on the right track here.  The one time that you'd really have to worry about doing the "return value mock" of a call with specific arguments (i.e., `.get(1).return_value`) is when the function call has some unique kind of chaining.  What I mean here is a function call where it only occurs once and in a very particular way.  For instance, maybe you would need to mock the return value of `get(1).deserialize().to_json()`; in that case, you'd need to supply that specific argument for the `get`.  Otherwise, if the `get` the last chained method call, just deal with the return value and don't worry about the arguments passed.

File: tests/test_interface.py [hunk 23]
```diff
@@ -399,6 +415,28 @@ def test_interface_delete_payment_with_invalid_id(self, mock_db):
         self.assertNotEqual(payment_after_deletion.id,invalid_id)
         #TODO - next sprint : do get, check count, should be same
 
+#valid PUT data
+    @mock.patch.object(PaymentService,'is_valid_put',return_value = True)
+    @mock.patch.object(app_db, 'session')
+    @mock.patch('app.db.models.Payment')
+    #@mock.patch.object(Payment, 'deserialize_put')
+    #@mock.patch.object(Payment, 'serialize', return_value=PUT_CREDIT_RETURN)
+    def test_interface_update_with_valid_put_data(self,mock_isvalid,mock_db,mock_P):
```
Comment (jdgumz) [inline tests/test_interface.py]: The order of the mocks here is incorrect - `mock_isvalid` should be the third of the three mock parameters, not the first.  See Derek's comment below for a correct ordering.

File: tests/test_interface.py (hunk 19 above)
Comment (jdgumz) [inline tests/test_interface.py]: As a quick addendum, we might want the assertion on `serialize` calls to be `assert mock_P.serialized.call_count == 2`.

**Comment by odoric27** [comment]
cool. i was wondering if there was a way to check in ways other than
`called_once` and `called`

File: app/payments.py (hunk 7 above)
Comment (SharanyaT) [inline app/payments.py]: updated - works fine

File: app/db/interface.py (hunk 22 above)
Comment (SharanyaT) [inline app/db/interface.py]: Sounds good - is_default,is_removed,charge_history will not be part of payment_replacement. user_id will be a part of the input (similar to POST)
Also, there is a separate ticket #135 - for test cases for is_valid_put method. I will check if the request data has any NON UPDATABLE FIELDS, if so, it will raise a data validation error.

Comment (SharanyaT) [inline app/db/interface.py]: For now, is_valid_put takes 2 parameters - request data and user id. User id is needed to validate if there are any changes to user_id field and raises an error acc. Test cases for is_valid will be handled in another PR as mentioned above

File: tests/test_interface.py (hunk 23 above)
Comment (SharanyaT) [inline tests/test_interface.py]: Fixed this

File: app/error_handlers.py (hunk 12 above)
Comment (SharanyaT) [inline app/error_handlers.py]: Fixed this up and renamed to InvalidPaymentIDError

File: app/error_handlers.py (hunk 16 above)
Comment (SharanyaT) [inline app/error_handlers.py]: Fixed it
